Mathlib/Topology/Algebra/Order/LiminfLimsup.lean

```diff
@@ -302,106 +302,143 @@ end LiminfLimsup
+    (f_decr : Antitone f) (f_cont : ContinuousAt f F.limsSup)
+    (bdd_above : F.IsBounded (· ≤ ·) := by isBoundedDefault)
+    (bdd_below : F.IsBounded (· ≥ ·) := by isBoundedDefault) :
+    f F.limsSup = F.liminf f := by
+  have cobdd : F.IsCobounded (· ≤ ·) := bdd_below.isCobounded_flip
   apply le_antisymm
-  · have A : { a : R | ∀ᶠ n : R in F, n ≤ a }.Nonempty := ⟨⊤, by simp⟩
-    rw [limsSup, f_decr.map_sInf_of_continuousAt' f_cont A]
+  · rw [limsSup, f_decr.map_sInf_of_continuousAt' f_cont bdd_above cobdd]
     apply le_of_forall_lt
     intro c hc
-    simp only [liminf, limsInf, lt_sSup_iff, eventually_map, Set.mem_setOf_eq, exists_prop,
-      Set.mem_image, exists_exists_and_eq_and] at hc ⊢
-    rcases hc with ⟨d, hd, h'd⟩
-    refine' ⟨f d, _, h'd⟩
-    filter_upwards [hd]with x hx using f_decr hx
-  · rcases eq_or_lt_of_le (bot_le : ⊥ ≤ F.limsSup) with (h | limsSup_ne_bot)
-    · rw [← h]
-      apply liminf_le_of_frequently_le _
-      apply frequently_of_forall
-      intro x
-      exact f_decr bot_le
-    by_cases h' : ∃ c, c < F.limsSup ∧ Set.Ioo c F.limsSup = ∅
+    simp only [liminf, limsInf, eventually_map] at hc ⊢
+    obtain ⟨d, hd, h'd⟩ :=
+      exists_lt_of_lt_csSup ((@Set.nonempty_image_iff R S f _).mpr bdd_above) hc
+    apply lt_csSup_of_lt ?_ ?_ h'd
+    · exact (Antitone.isBoundedUnder_le_comp f_decr bdd_below).isCoboundedUnder_flip
+    · rcases hd with ⟨e, ⟨he, fe_eq_d⟩⟩
+      filter_upwards [he] with x hx using (fe_eq_d.symm ▸ f_decr hx)
+  · by_cases h' : ∃ c, c < F.limsSup ∧ Set.Ioo c F.limsSup = ∅
     · rcases h' with ⟨c, c_lt, hc⟩
       have B : ∃ᶠ n in F, F.limsSup ≤ n := by
-        apply (frequently_lt_of_lt_limsSup (by isBoundedDefault) c_lt).mono
+        apply (frequently_lt_of_lt_limsSup cobdd c_lt).mono
         intro x hx
         by_contra'
         have : (Set.Ioo c F.limsSup).Nonempty := ⟨x, ⟨hx, this⟩⟩
         simp only [hc, Set.not_nonempty_empty] at this
-      apply liminf_le_of_frequently_le _
-      exact B.mono fun x hx ↦ f_decr hx
+      apply @liminf_le_of_frequently_le R S _ F f (f (limsSup F)) (B.mono fun x hx ↦ f_decr hx) ?_
+      obtain ⟨b, hb⟩ := bdd_above
+      use f b
+      simp only [ge_iff_le, eventually_map]
+      filter_upwards [hb] with t ht using f_decr ht
+    simp only [gt_iff_lt, not_lt, ge_iff_le, not_exists, not_and] at h'
     by_contra' H
     obtain ⟨l, l_lt, h'l⟩ : ∃ l < F.limsSup, Set.Ioc l F.limsSup ⊆ { x : R | f x < F.liminf f }
-    exact exists_Ioc_subset_of_mem_nhds ((tendsto_order.1 f_cont.tendsto).2 _ H) ⟨⊥, limsSup_ne_bot⟩
+    · apply exists_Ioc_subset_of_mem_nhds ((tendsto_order.1 f_cont.tendsto).2 _ H)
+      by_contra con
+      simp only [not_exists, not_lt] at con
+      simpa only [lt_self_iff_false] using lt_of_le_of_lt
+        (@liminf_le_of_frequently_le R S _ F f (f (limsSup F))
+          (frequently_of_forall (fun r ↦ f_decr (con r))) (bdd_above.isBoundedUnder f_decr)) H
```

Comment:
Sorry, I honestly can't form an opinion about that!

I guess I mainly opted for not talking about a bottom element, since under the conditionally complete hypotheses it is not given that a bottom element exists. But the proof ends up being a bit of a case bash anyway, and one way to handle some cases would be to consider the case that there indeed exists a bottom element and `F.limsSup` happens to be that.

Is it the number of (nested) reasoning steps that argue by contradiction that seems questionable style here?

I really don't know what would be elegant (or the least inelegant) here. Do you have a clear opinion?

Reply:
Thinking about it again I think your structure is actually good (and there was no style issue at all with that, I was mostly nit-picking). The only thing I'd change about the structure is I think it would be clearer if you did 
```suggestion
have : \not IsBot F.limsSup := by sorry
```
after `by_contra' H` and then use that in the obtain. Is that clear enough?